MAINT Common sample_weight validation

rth opened the pull request:
This implements a helper function for sample weight validation, and uses it consistently across the code base when applicable.

Currently it's only applied to estimators (particularly in linear models and SVM) that already had some form of sample weight validation. It the future it might be worthwhile to add it to estimators that don't currently validate sample weights.

It is a pre-requisite for optionally enforcing positive sample weights (https://github.com/scikit-learn/scikit-learn/issues/12464)

Also somewhat related to https://github.com/scikit-learn/scikit-learn/pull/14286 and https://github.com/scikit-learn/scikit-learn/pull/14294

The main motivation is to avoid implementing this yet another time for GLMs in https://github.com/scikit-learn/scikit-learn/pull/14300

Any thought on this @glemaitre , since you were looking into related issues lately?

rth committed:
Common sample_weight validation in huber

rth committed:
Use _check_sample_weight in linear_model/stochastic_gradient

rth committed:
Use check_sample weight in logistic

rth committed:
Better order parameter

rth committed:
Sample weight checks in svm

rth committed:
Tests

rth commented on sklearn/linear_model/huber.py:
```diff
@@ -253,11 +253,8 @@ def fit(self, X, y, sample_weight=None):
-            sample_weight = np.array(sample_weight)
-            check_consistent_length(y, sample_weight)
-        else:
-            sample_weight = np.ones_like(y)
```
I'm pretty sure this didn't produce sample weights with the expected dtype when `y` was an integer array.

rth commented on sklearn/utils/validation.py:
```diff
@@ -980,3 +980,56 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+    if n_samples is not None and y is not None:
+        raise ValueError('Only one of y, n_samples must be provided!')
+    elif y is not None:
+        n_samples = y.shape[0]
```
I think should handle both the cases when,
 - y is 1D
 - y is 2D (a 1D column)
 - y is 2D (multi-output)

rth committed:
Minor fix

rth committed:
Better handle numeric sample_weight

jeremiedbb commented on sklearn/utils/validation.py:
```diff
@@ -980,3 +980,60 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+        n_samples = y.shape[0]
+
+    if sample_weight is None or isinstance(sample_weight, numbers.Number):
+        if order is None:
```
I don't see any reason for `sample_weight` to be a number. It's equivalent to all ones.
Only Ridge allows `sample_weight` to be a float, but I think it's useless and we should rather not allow a float as `sample_weight` and not support it here.

rth replied:
I also don't see the point of it, but this PR aims to do a refactoring without breaking backward compatibility. If we are going to disallow it (and potentially break some user code) I would rather it be considered in a separate PR.

jeremiedbb commented on sklearn/utils/validation.py:
```diff
@@ -980,3 +980,60 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+
+
+def _check_sample_weight(sample_weight, y=None, n_samples=None, dtype=None,
+                         order=None, **kwargs):
```
some estimators (e.g. AdaBoost) might need the weights to be normalized (i.e. `w / w.sum()`). You could add a normalize parameter ?

glemaitre replied:
Or do we let each estimator do it since that it might be on estimator bases?

rth replied:
Yeah, for now the approach was to add functionality as I encountered it in the code. `AdaBoost` seems to do something fairly specific, I haven't seen the need for it in other estimators. But if we encounter more, we could certainly add it. This would need to be applied to more estimators in the future (and that's also why I would rather not make the function public, as parameters may change)..

jeremiedbb replied:
I agree to make it private until all estimators use it and the functionalities are stable.

jeremiedbb replied:
It seem to be specific to AdaBoost. Kmeans uses a different normalization (weights have to sum to n_samples...)

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -980,3 +980,56 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+       target variable. Either y or n_samples must be provided.
+    n_samples: int or None
+       expected number of samples. Either y or n_samples must be provided.
+    dtype: dtype
```
```suggestion
    dtype : dtype
```

jeremiedbb commented on sklearn/utils/validation.py:
```diff
@@ -980,3 +980,60 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+       target variable. Either y or n_samples must be provided.
+    n_samples: int or None
+       expected number of samples. Either y or n_samples must be provided.
+    dtype: dtype
```
you need `n_samples` in case y isn't there for unsupervised ?

rth replied:
yes, e.g. k-means.

jeremiedbb replied:
should we than pass X instead of n_samples ?

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -980,3 +980,56 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+    n_samples: int or None
+       expected number of samples. Either y or n_samples must be provided.
+    dtype: dtype
+       dtype of the validated sample_weight
```
```suggestion
       dtype of the validated `sample_weight`.
```

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -980,3 +980,56 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+        otherwise it will be of 'C' order by default.
+
+    kwargs:
+       additional parameters to pass to check_array
```
```suggestion
       Additional parameters to pass to `check_array`
```

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -980,3 +980,56 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+        nothing is ensured about the memory layout of the output array,
+        otherwise it will be of 'C' order by default.
+
+    kwargs:
```
```suggestion
    kwargs :
```

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -980,3 +980,56 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+
+    Parameters
+    ----------
+    sample_weight : ndarray, shape=(n_samples,)
```
```suggestion
    sample_weight : ndarray, shape (n_samples,)
```

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -980,3 +980,56 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+    Parameters
+    ----------
+    sample_weight : ndarray, shape=(n_samples,)
+       validated sample weights
```
```suggestion
       Validated sample weights.
```

jeremiedbb commented on sklearn/utils/validation.py:
```diff
@@ -980,3 +980,60 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+    else:
+        if dtype is None:
+            dtype = [np.float64, np.float32]
+        sample_weight = check_array(
```
Are there cases where we don't want `sample_weight.dtype` to be the same as `X.dtype` ?
Because if there aren't any we can pass X as a parameter, force X.dtype and get rid of the `n_samples` parameter.

rth replied:
> Are there cases where we don't want sample_weight.dtype to be the same as X.dtype ?

That's a very good question, but I'm not 100% sure where `sample_weight` are used across estimators. For instance, take the case or some regressor where `X` is `float32`, and `y` is `float64`. Right now `sample_weight` would be `float64`  by default. Would it make sense to also make it float32? I don't know, maybe. Would it break something, and do we currently have sufficient tests to detect it? I don't know either.

So here I went with the safer road of just adding this refactoring without changing dtypes. The second part could be added but I think it would deserve a more careful look as part of https://github.com/scikit-learn/scikit-learn/issues/11000 etc.

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -980,3 +980,60 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+                sample_weight, accept_sparse=False,
+                ensure_2d=False, dtype=dtype, order=order, **kwargs
+        )
+        if sample_weight.ndim != 1:
```
you can use `column_or_1d` validation to have consistent warning and error raised

rth replied:
So my problem with `column_or_1d` is that applies `np.ravel` even to 1D arrays, making the provided `order` irrelevant as the output is then contiguous (for the cost of a copy).

Also it doesn't look like this unraveling was previously done for sample_weight in the estimators I looked at. Meaning that 1D column like `sample_weight` I imagine would previously fail, and with this would pass. Do we really want to add this additional degree or freedom?

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -980,3 +980,60 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+        if sample_weight.ndim != 1:
+            raise ValueError("Sample weights must be 1D array or scalar")
+
+        if sample_weight.shape != (n_samples,):
```
you can use `check_consistency_length`

glemaitre replied:
to have consistent error raised

rth replied:
Yes, but `y` is not always provided, and `check_consistency_length` wouldn't work when only `n_samples` is provided.

glemaitre replied:
uhm True. It is a pity that we cannot keep the error message in one location.

jeremiedbb replied:
unless you check against X :)

rth replied:
You are right, checking against `X` is probably best. Will add it.

thomasjpfan replied:
Was this addressed?

rth replied:
No. I just want to check the shape, while `check_consistent_length` does a bunch of generic and more complicated stuff. I'm also not convinced that the resulting generic error message,
```
Found input variables with inconsistent numbers of samples: %r
```
is an improvement over the current specific message,
```
samples_weight.shape == {}, expected {}!
```
(though I'm open to suggestions on how to improve the latter one).

rth replied:
The difference is also that by this point `sample_weight` has been validated, so I am sure it's an array, while `check_consistent_length` doesn't make any assumptions hence the more complicated and generic solution.

thomasjpfan replied:
Makes sense.

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -980,3 +980,56 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+                         order=None, **kwargs):
+    """Validate sample weights
+
+    Parameters
```
We can introduce this in the next PR if we don't break any back-compatibility.

rth replied:
Yes, I would rather make a separate PR for, it might also involve defining a global setting `require_positive_sample_weight` to ether 'all', 'any' or False cf related discussion in https://github.com/scikit-learn/scikit-learn/issues/12464#issuecomment-433815773

glemaitre commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -853,3 +855,28 @@ def test_check_scalar_invalid(x, target_name, target_type, min_val, max_val,
+
+
+def test_check_sample_weight():
+    with pytest.raises(ValueError,
```
Do you want to group all the exception within the same parametrize test ;)

rth replied:
There is only two exceptions here, and the parameters used are not the same, so I'm not convinced that parametrization would improve readability..

glemaitre commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -853,3 +855,28 @@ def test_check_scalar_invalid(x, target_name, target_type, min_val, max_val,
     assert type(raised_error.value) == type(err_msg)
+
+
+def test_check_sample_weight():
```
I think that we need additional tests to ensure that we pass correctly the `kwargs` to `check_array`

rth replied:
Actually kwargs are not longer needed, so I remove them. They can always be added later if necessary.

rth commented on sklearn/utils/validation.py:
```diff
@@ -980,3 +980,56 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
         raise ValueError('`{}`= {}, must be <= {}.'.format(name, x, max_val))
+
+
+def _check_sample_weight(sample_weight, y=None, n_samples=None, dtype=None,
```
I don't think we should maintain backward compatibility on this or lets users use it in their code. Utils are semi-private anyway, this just marks it as explicitly private. We can still use it in common checks, even if it's private right?

rth committed:
Update sklearn/utils/validation.py

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

rth committed:
Update sklearn/utils/validation.py

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

rth committed:
Update sklearn/utils/validation.py

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

rth committed:
Update sklearn/utils/validation.py

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

rth committed:
Update sklearn/utils/validation.py

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

rth committed:
Update sklearn/utils/validation.py

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

rth committed:
Update sklearn/utils/validation.py

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

rth committed:
Update sklearn/utils/validation.py

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

rth committed:
Update sklearn/utils/validation.py

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

rth committed:
Update sklearn/utils/validation.py

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

rth committed:
Remove kwargs

rth commented:
Thanks for reviews @glemaitre and @jeremiedbb . I think I addressed all comments. Please let me know if I missed something.

>> unless you check against X :)
>
> You are right, checking against X is probably best. Will add it.

Actually I have looked into it, and it's not that simple. Checking against `X` works well if `check_array` was run on `X` and we are sure it's an array like. However there is no guarantee that this happened when we call `_check_sample_weight` in general.

In particular, `X` could be any iterable, that upon `check_array(X)` would return an array. In which case we don't know the number of samples. Of course we can check that it's an ndarray or sparse array and raise an error otherwise, but again there is no guarantee that we have tests that will cover this use case for the particular estimator that may have this issue.

I guess one could make an argument that `y` could also be an iterable, but here I used it when it was used before, so it's a net zero change in that respect.

In the end, although manually passing `n_samples` or `y` may require a tiny bit more of work, in the end I prefer it as it's safer (and I don't want to deal with all the edge cases that may arise when checking against `X`).

jeremiedbb commented:
My last argument:
you can get n_samples from `_num_samples(X)`, and if X is such that this would raise an error, it would have been caught later by check_array anyway (all estimators call check_array at some point)
But if you think it's safer this way, let it as is.

rth committed:
Use _check_sample_weight(sample_weight, X)

rth committed:
Update kmeans to use _check_sample_weight

rth commented:
> you can get n_samples from _num_samples(X), and if X is such that this would raise an error,

OK, you convinced me :). Changed it to `_check_sample_weight(sample_weight, X)`.

Also as you mentioned there was an identically named function in kmeans, that also did normalization. Changed it to `_check_normalize_sample_weight` and made it use this helper.

rth commented:
> Are there cases where we don't want sample_weight.dtype to be the same as X.dtype ?

Also `X` can be int, that passes through `check_array` unchanged, and we definitely don't want `sample_weight` to be int.  Looks like this was happening in kmeans before and wasn't caught, and also in all other places that used `dtype=X.dtype`. Working on a fix.

rth committed:
Non float dtype should not be supported

rth committed:
Merge remote-tracking branch 'upstream/master' into check-sample-weight

jeremiedbb commented on sklearn/cluster/k_means_.py:
```diff
@@ -164,19 +164,19 @@ def _tolerance(X, tol):
+    sample_weight_was_none = sample_weight is None
+
+    sample_weight = _check_sample_weight(sample_weight, X, dtype=X.dtype)
+    if not sample_weight_was_none:
```
I don't think this `sample_weight_was_none` thing is necessary.
```python
sample_weight = _check_sample_weight(sample_weight, X, dtype=X.dtype)
if sample_weight.sum() != 1:
    # normalize the weights to sum up to n_samples
    n_samples = len(sample_weight)
    scale = n_samples / sample_weight.sum()
    sample_weight *= scale
return sample_weight
```
should be enough

jeremiedbb replied:
edited

rth replied:
Yeah, but we would be computing one extra sum with the above, also float equality comparison is not that sure. Another solution could be,
```py
sample_weight_validated = _check_sample_weight(sample_weight, X, dtype=X.dtype)
if sample_weight is None:
    # normalize the weights to sum up to n_samples
    scale = n_samples / sample_weight.sum()
    sample_weight_validated *= scale
return sample_weight_validated
```
I can change for it if you prefer. Either way defining one extra temp variable is necessary..

jeremiedbb replied:
You're right, there's no need to compute a sum when not necessary.

> I can change for it if you prefer.

no need

jeremiedbb commented on sklearn/utils/validation.py:
```diff
@@ -980,3 +980,56 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+    sample_weight : {ndarray, Number or None}, shape (n_samples,)
+       Input sample weights.
+    X : nd-array, list or sparse matrix
+        Input data.
```
There should be a blank line between parameters, right ?

jeremiedbb commented on sklearn/utils/validation.py:
```diff
@@ -980,3 +980,56 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+        otherwise it will be of 'C' order by default.
+
+    Parameters
+    ----------
```
Returns

rth commented:
I have added a fix so that when `dtype` is not float (e.g. an int), the resulting sample_weight will have the default float dtype (e.g. float64) see https://github.com/scikit-learn/scikit-learn/issues/14314 for mode details.

CI is green now. Any other comments that I should address?

jeremiedbb commented on sklearn/utils/validation.py:
```diff
@@ -980,3 +980,56 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+       dtype of the validated `sample_weight`. Note that if `dtype` is not
+       one of `float32`, `float64`, the output will be of dtype `float64`.
+    order : 'F', 'C' or None (default=None)
+        Whether an array will be forced to be fortran or c-style.
```
'F' is equivalent to 'C' for 1D array. Don't you think we can drop this parameter and always enforce 'C'.
Letting non contiguous array pass is risky if it's passed to cython, with pointers manipulation

rth replied:
I agree.

rth committed:
Drop order parameter + address review comments

rth committed:
Fix typo

rth committed:
Fix tests

rth commented:
Thanks @jeremiedbb . I addressed your above comments.

jeremiedbb commented on sklearn/linear_model/ridge.py:
```diff
@@ -428,8 +429,7 @@ def _ridge_regression(X, y, alpha, sample_weight=None, solver='auto',
     if has_sw:
-        if np.atleast_1d(sample_weight).ndim > 1:
-            raise ValueError("Sample weights must be 1D array or scalar")
+        sample_weight = _check_sample_weight(sample_weight, y)
```
it's probably equivalent but y -> X ?

jeremiedbb reviewed (approved):
LGTM !

rth committed:
Address review

thomasjpfan commented on sklearn/utils/validation.py:
```diff
@@ -980,3 +980,50 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+    """
+    n_samples = _num_samples(X)
+
+    if dtype is not None and dtype not in [np.float32, np.float64]:
```
Test for when `dtype=np.float32` the `sample_weight` is `np.float32`

rth replied:
Thanks @thomasjpfan, added the test.

rth committed:
Attempt to fix 32bit / 64bit comparison test

rth committed:
Check that float32 dtype is preserved

rth commented:
@glemaitre , @thomasjpfan I think I have addressed all of your comments in this PR.

thomasjpfan commented on sklearn/utils/validation.py:
```diff
@@ -980,3 +980,56 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+
+def _check_sample_weight(sample_weight, y=None, n_samples=None, dtype=None,
+                         order=None, **kwargs):
+    """Validate sample weights
```
I can't press the "Commit suggestion" button. :(

rth replied:
Done missed that one. Thanks!

rth committed:
Update sklearn/utils/validation.py

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

rth committed:
Lint

NicolasHug commented on sklearn/utils/validation.py:
```diff
@@ -980,3 +980,50 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+            raise ValueError("Sample weights must be 1D array or scalar")
+
+        if sample_weight.shape != (n_samples,):
+            raise ValueError("samples_weight.shape == {}, expected {}!"
```
```suggestion
            raise ValueError("sample_weight.shape == {}, expected {}!"
```

NicolasHug reviewed (commented):
Nitpicks + 1 small concern

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -980,3 +980,50 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+
+    if sample_weight is None or isinstance(sample_weight, numbers.Number):
+        if sample_weight is None:
+            sample_weight = np.ones(n_samples, dtype=dtype)
```
Uhm good to know. Up to know, lots of estimators used this as an initialization. Would this require a parameter?

rth replied:
It's a good remark. So far, this function is a way of factorizing code in linear models that do this a lot. If weights are assumed to be an array later on, frequently it's better to convert None to `np.ones` earlier than later.

For instance an antypattern example is logistic regression that before this PR used to pass `sample_weight=None` though, and then allocate an array of ones repeatedly during [each call of the loss function](https://github.com/scikit-learn/scikit-learn/blob/f82d966d7b6286aa34a11374304c0bca515f4331/sklearn/linear_model/logistic.py#L161).

For estimators that don't use sample weight at all, the easiest is to not validate them if they are None,
```
if sample_weight is not None:
   sample_weight = _check_sample_weight(...)
```
IMO that's simpler and more readable  than adding a parameter `none_to_ones=False` (or something similar to the validation function).

glemaitre replied:
I like this solution. @NicolasHug do you have the same feeling

NicolasHug replied:
Me too, thanks!

rth committed:
Update sklearn/utils/validation.py

Co-Authored-By: Nicolas Hug <contact@nicolas-hug.com>

rth committed:
Improve docstring wording

rth commented on sklearn/utils/validation.py:
```diff
@@ -980,3 +980,50 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
+
+    dtype: dtype
+       dtype of the validated `sample_weight`. Note that if `dtype` is not
+       one of `float32`, `float64`, the output will be of dtype `float64`.
```
Indeed. Improved the wording.

rth committed:
Fix tests

rth commented:
Thanks for the review @NicolasHug! any other comments ? Or can this be merged?

NicolasHug commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -853,3 +855,35 @@ def test_check_scalar_invalid(x, target_name, target_type, min_val, max_val,
+
+    # check numbers input
+    sample_weight = _check_sample_weight(2.0, X=np.ones((5, 2)))
+    assert_allclose(sample_weight, 2*np.ones(5))
```
Weird that we don't get a flake8 warning here

NicolasHug commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -853,3 +855,35 @@ def test_check_scalar_invalid(x, target_name, target_type, min_val, max_val,
+    # int dtype will be converted to float64 instead
+    X = np.ones((5, 2), dtype=np.int)
+    sample_weight = _check_sample_weight(None, X, dtype=X.dtype)
+    assert sample_weight.dtype == np.float64
```
Maybe add a test for the shape difference error? (it's implicitly tested in kmeans)

NicolasHug reviewed (approved):
2 nitpicks but LGTM, feel free to merge

rth committed:
Nicolas's comments

rth commented:
Thanks @NicolasHug , I addressed your comments.

thomasjpfan commented:
Thank you @rth!
